### PR TITLE
fix: handle comma-separated provided IPs

### PR DESCRIPTION
**Pull request by mainred**
#### What type of PR is this?
/kind bug
For dual-stack Windows case, both ipv4 and ipv6 ip address are provided and separated by comma in a string. This change is to split the ipv4 and ipv6 addresses and check whether all these provided IPs exist in the cloud provider.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MartinForReal** [comment]
Could you please update the release note?

File: pkg/nodemanager/nodemanager.go [hunk 1]
```diff
@@ -602,19 +598,33 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
...
+	if !ok || len(providedIPStr) == 0 {
+		return []v1.NodeAddress{}, true
+	}
+	var existNodeIPs []v1.NodeAddress
+	nodeIPsExists := false
+	providedIPs := strings.Split(providedIPStr, ",")
+	for i := range nodeAddresses {
+		for _, providedIP := range providedIPs {
+			fmt.Println(nodeAddresses[i].Address, providedIP)
+
```
Comment (feiskyer) [inline pkg/nodemanager/nodemanager.go:615]: could you trim spaces of providedIP before compare?

File: pkg/nodemanager/nodemanager.go [hunk 2]
```diff
@@ -602,19 +598,33 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
...
+	providedIPStr, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]
+	if !ok || len(providedIPStr) == 0 {
+		return []v1.NodeAddress{}, true
+	}
+	var existNodeIPs []v1.NodeAddress
+	nodeIPsExists := false
+	providedIPs := strings.Split(providedIPStr, ",")
+	for i := range nodeAddresses {
+		for _, providedIP := range providedIPs {
+			fmt.Println(nodeAddresses[i].Address, providedIP)
```
Comment (feiskyer) [inline pkg/nodemanager/nodemanager.go:614]: remove L614 and L617. we should not println directly

File: pkg/nodemanager/nodemanager.go [hunk 3]
```diff
@@ -602,19 +598,33 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
...
 }
 
-func ensureNodeProvidedIPExists(node *v1.Node, nodeAddresses []v1.NodeAddress) (*v1.NodeAddress, bool) {
-	var nodeIP *v1.NodeAddress
-	nodeIPExists := false
-	if providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]; ok {
-		nodeIPExists = true
-		for i := range nodeAddresses {
+// Ensure all provided node ip addresses are found in the cloud provider, otherwise return false
+// and the existing node addresses.
```
Comment (feiskyer) [inline pkg/nodemanager/nodemanager.go:602]: L602 is not needed

File: pkg/nodemanager/nodemanager.go [hunk 4]
```diff
@@ -302,10 +302,8 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
 	}
 	// If nodeIP was suggested by user, ensure that
 	// it can be found in the cloud as well (consistent with the behaviour in kubelet)
-	if nodeIP, ok := ensureNodeProvidedIPExists(node, nodeAddresses); ok {
-		if nodeIP == nil {
-			return fmt.Errorf("specified Node IP %s not found in cloudprovider for node %q", nodeAddresses, node.Name)
-		}
+	if existNodeIPs, ok := ensureNodeProvidedIPsExists(node, nodeAddresses); !ok {
+		return fmt.Errorf("not all specified Node IP %s not found in cloudprovider for node %q, existing Node IPs are %s ", nodeAddresses, node.Name, existNodeIPs)
```
Comment (feiskyer) [inline pkg/nodemanager/nodemanager.go:306]: nit: remove the second not

File: pkg/nodemanager/nodemanager.go (hunk 1 above)
Comment (mainred) [reply pkg/nodemanager/nodemanager.go]: Good point, I detected this in the UT as well. wanted to ask whether it's necessary or not.
Included an UT to test extra space as well.

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**